instrumentation/testutils/testtracer/testtracer.go

```diff
@@ -0,0 +1,272 @@
+	Service    string             `json:"service"`
+	Resource   string             `json:"resource"`
+	Type       string             `json:"type"`
+	Start      int64              `json:"start"`
+	Duration   int64              `json:"duration"`
+	Meta       map[string]string  `json:"meta"`
+	MetaStruct map[string]any     `json:"meta_struct"`
+	Metrics    map[string]float64 `json:"metrics"`
+	SpanID     uint64             `json:"span_id"`
+	TraceID    uint64             `json:"trace_id"`
+	ParentID   uint64             `json:"parent_id"`
+	Error      int32              `json:"error"`
+	SpanLinks  []SpanLink         `json:"span_links"`
+}
+
+// SpanLink defines a span link with the same format as it is sent to the agent.
+type SpanLink struct {
+	TraceID     uint64            `json:"trace_id"`
+	TraceIDHigh uint64            `json:"trace_id_high"`
+	SpanID      uint64            `json:"span_id"`
+	Attributes  map[string]string `json:"attributes"`
+	Tracestate  string            `json:"tracestate"`
+	Flags       uint32            `json:"flags"`
+}
+
+// TestTracer is an inspectable tracer useful for tests.
+type TestTracer struct {
+	Spans        <-chan Span
+	roundTripper *mockTransport
+}
+
+// Start calls [tracer.Start] with a mocked transport and provides a new [TestTracer] that allows to inspect
+// the spans produced by this application.
+func Start(t *testing.T, opts ...Option) TestTracer {
+	cfg := defaultConfig()
+	for _, opt := range opts {
+		opt(cfg)
+	}
+
+	spansChan := make(chan Span)
+	rt := &mockTransport{
+		T:         t,
+		spansChan: spansChan,
+		agentInfo: cfg.AgentInfoResponse,
+	}
+	httpClient := &http.Client{
+		Transport: rt,
+	}
+	tt := TestTracer{Spans: spansChan, roundTripper: rt}
+	t.Cleanup(tt.Stop)
+
+	startOpts := append([]tracer.StartOption{
+		tracer.WithEnv("TestTracer"),
+		tracer.WithService("TestTracer"),
+		tracer.WithServiceVersion("1.0.0"),
+		tracer.WithHTTPClient(httpClient),
+		tracer.WithLogger(&testLogger{T: t}),
+	}, cfg.TracerStartOpts...)
+
+	err := tracer.Start(startOpts...)
+	require.NoError(t, err)
+
+	return tt
+}
+
+type config struct {
+	TracerStartOpts   []tracer.StartOption
+	AgentInfoResponse AgentInfo
+}
+
+func defaultConfig() *config {
+	return &config{
+		TracerStartOpts:   nil,
+		AgentInfoResponse: AgentInfo{},
+	}
+}
+
+type Option func(*config)
+
+// WithTracerStartOpts allows to set [tracer.StartOption] on the tracer.
+func WithTracerStartOpts(opts ...tracer.StartOption) Option {
+	return func(cfg *config) {
+		cfg.TracerStartOpts = opts
+	}
+}
+
+// WithAgentInfoResponse sets a custom /info agent response. It can be used to enable/disable certain features
+// from the tracer that depend on whether the agent supports them or not.
+func WithAgentInfoResponse(response AgentInfo) Option {
+	return func(cfg *config) {
+		cfg.AgentInfoResponse = response
+	}
+}
+
+// Stop stops the tracer. It should be called after the test finishes.
+func (tt TestTracer) Stop() {
```

Comment:
Shouldn't this receiver here (and in WaitForSpans) be a pointer receiver? We are copying a channel on every call.